### PR TITLE
Dev implement imu sensors

### LINKED ISSUE #248: Implement IMU sensors

### PR DESCRIPTION
**Pull request by ThorsHamster**
Closes #248 

Adds Classes for IMU sensors with sensitivity (for Gyroscope and Accelerometer) and without Sensitivity (Magnetometer).
Adds also first draft of Class for Gyroscope.

Next steps will be (own Pullrequests):
* Add of MPU9255 Class
* Initialization of all Sensors (Gyroscope, Accelerometer, Magnetometer) inside of specific classes and MPU9255 Class.
* Selftest of sensors
* Calculate mean value of sensors
* Calibration of sensors
* Calculation of quaternions

Here you can see the output over UART with original MPU9255:
Measurement_IMU

Keep in mind: The sensors are not calibrated or selftested. The values are fluctuating a lot.

### CODE REVIEW

File: src/imu/imu_sensor.hpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  auto ReadContentFromRegister(std::uint8_t read_from_register, std::uint16_t byte_size) noexcept -> std::vector<std::uint8_t>;
+  auto ImuConnectionSuccessful(void) noexcept -> bool;
+  auto ImuConnectionFailed(void) noexcept -> bool;
+  auto SetI2CAdress(std::uint8_t i2c_address) noexcept -> void;
+  auto ConvertUint8BytesIntoInt16SensorValue(std::uint8_t first_byte, std::uint8_t second_byte) noexcept -> std::int16_t;
+  auto IsInitialized(void) noexcept -> bool;
+  auto SetSensorValues(std::int16_t x, std::int16_t y, std::int16_t z) noexcept -> void;
+
+  types::EuclideanVector<std::int16_t> sensor_values_{-1, -1, -1};
+  std::unique_ptr<i2c::I2CInterface> i2c_handler_;
```
Comment (elheck) [inline src/imu/imu_sensor.hpp:34]: Instance fields/attributes are protected, but in the class diagram the are indicated to be private (by the - , # is protected)

File: src/imu/imu_sensor.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+class InertialMeasurementSensor {
+ public:
+  virtual ~InertialMeasurementSensor() = default;
+
+  explicit InertialMeasurementSensor(std::unique_ptr<i2c::I2CInterface> i2c_handler) : i2c_handler_(std::move(i2c_handler)){};
+  auto Get(void) noexcept -> types::EuclideanVector<int16_t>;
+  auto Update(void) noexcept -> types::DriverStatus;
+
+ protected:
+  auto Mpu9255Detected(void) noexcept -> bool;
```
Comment (elheck) [inline src/imu/imu_sensor.hpp:21]: why are these methods protected and non-virtual. Non-virtual indicated that you don't want them overwritten, by an inheriting class. Non virtual changes the look up mechanism of c++ when inheriting from this class.

File: src/imu/imu_sensor.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
+#include "imu_sensor.hpp"
+
+namespace imu {
+
+auto InertialMeasurementSensor::Update(void) noexcept -> types::DriverStatus {
+  if (!IsInitialized())
+    return types::DriverStatus::HAL_ERROR;
+
+  std::vector<uint8_t> measurement_values;
```
Comment (elheck) [inline src/imu/imu_sensor.cpp]: you could combine these two lines to auto measurement_values = ReadContentFromRegister...

File: src/imu/imu_sensor.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  measurement_values = ReadContentFromRegister(SENSOR_DATA_REGISTER, 6);
```
Comment (elheck) [inline src/imu/imu_sensor.cpp]: the number 6 is a unnamed magic number, what does it mean. Also it is an implicit conversion from int to unit16_t

File: src/imu/imu_sensor.cpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+  if (ImuConnectionSuccessful()) {
+    SetSensorValues(
+        ConvertUint8BytesIntoInt16SensorValue(measurement_values.at(0), measurement_values.at(1)),
```
Comment (elheck) [inline src/imu/imu_sensor.cpp:14]: Magic numbers. You could use a private enum to name the positions of the measurement values

File: src/imu/imu_sensor.cpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+  return types::DriverStatus::OK;
+}
+
+auto InertialMeasurementSensor::Get(void) noexcept -> types::EuclideanVector<int16_t> {
+  return sensor_values_;
+}
+
+auto InertialMeasurementSensor::Mpu9255Detected(void) noexcept -> bool {
+  bool who_i_am_correct = false;
```
Comment (elheck) [inline src/imu/imu_sensor.cpp]: The i2c register for selfidentification is commonly named "who am i" not "who i am"

File: src/imu/imu_sensor.cpp [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+auto InertialMeasurementSensor::Write(const std::vector<std::uint8_t>& data) noexcept -> void {
+  imu_status_ = i2c_handler_->Write(i2c_address_, data);
+}
+
+auto InertialMeasurementSensor::ImuConnectionSuccessful(void) noexcept -> bool {
+  return imu_status_ == types::DriverStatus::OK;
+}
+
+auto InertialMeasurementSensor::ImuConnectionFailed(void) noexcept -> bool {
```
Comment (elheck) [inline src/imu/imu_sensor.cpp:61]: Is this duplicated code? It is !ImuConnectionSuccessful

File: src/imu/imu_sensor_with_sensitivity.cpp [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+  sensitivity_ = sensitivity;
+}
+
+auto InertialMeasurementSensorWithSensitivity::SendSensitivityRegisterData(types::ImuSensitivity sensitivity) noexcept -> void {
+  std::uint8_t new_config = GetConfigRegisterDataForSensitivity(sensitivity);
+  Write({CONFIG_REGISTER, new_config});
+}
+
+auto InertialMeasurementSensorWithSensitivity::GetConfigRegisterDataForSensitivity(types::ImuSensitivity sensitivity) noexcept -> std::uint8_t {
+  std::uint8_t fs_sel = 0;
```
Comment (elheck) [inline src/imu/imu_sensor_with_sensitivity.cpp]: fs_sel? What does that mean? It is not a self explaining name

File: src/imu/imu_sensor_with_sensitivity.cpp [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+  } else if (sensitivity == types::ImuSensitivity::ROUGHER) {
+    fs_sel = imu::ImuSensitivityFSSelect::IMU_FS_ROUGHER;
+  } else if (sensitivity == types::ImuSensitivity::ROUGHEST) {
+    fs_sel = imu::ImuSensitivityFSSelect::IMU_FS_ROUGHEST;
+  }
+
+  std::vector<uint8_t> config_data;
+  config_data = ReadContentFromRegister(CONFIG_REGISTER, 1);
+
+  return static_cast<std::uint8_t>(config_data.at(0) & (0b11100111 | (fs_sel << 3)));
```
Comment (elheck) [inline src/imu/imu_sensor_with_sensitivity.cpp]: what is 0b11100111 and why do you shift fs_sel by three. This is hard to follow, maybe exchange the magic numbers by named variables. It might become more clear by doing this.

File: src/imu/mpu9255_data.hpp [hunk 10]
```diff
@@ -0,0 +1,16 @@
+#ifndef SRC_MPU9255_REGISTERS_HPP_
+#define SRC_MPU9255_REGISTERS_HPP_
+
+#include <cstdint>
+
+namespace imu {
+
+static constexpr std::uint8_t GYRO_CONFIG = 0x1B;
+static constexpr std::uint8_t GYRO_XOUT_H = 0x43;
```
Comment (elheck) [inline src/imu/mpu9255_data.hpp:9]: What does XOUT_H mean

File: tests/mock_libraries/imu/mock_i2c.hpp [hunk 11]
```diff
@@ -0,0 +1,14 @@
+#include <gmock/gmock.h>
+#include <memory>
+#include <tuple>
```
Comment (elheck) [inline tests/mock_libraries/imu/mock_i2c.hpp]: why do you need tuple

**Review by elheck — CHANGES_REQUESTED** [review]

File: src/imu/imu_sensor.hpp (hunk 1 above)
Comment (ThorsHamster) [reply src/imu/imu_sensor.hpp:34]: Thats correct. Fixed it in diagram.

File: src/imu/imu_sensor.hpp (hunk 2 above)
Comment (ThorsHamster) [reply src/imu/imu_sensor.hpp:21]: This is intended. The derived class should use the implementation of the base class. The derived class has just some functions added to the base class. (It is more specific).

File: src/imu/imu_sensor.cpp (hunk 3 above)
Comment (ThorsHamster) [reply src/imu/imu_sensor.cpp]: Correct. I fixed it. https://github.com/Squiddrone/Squiddrone/pull/268/commits/fd8c8b60ed066d257cce897a9452fec4224213f4

File: src/imu/imu_sensor.cpp (hunk 4 above)
Comment (ThorsHamster) [reply src/imu/imu_sensor.cpp]: Correct. Fixed it. https://github.com/Squiddrone/Squiddrone/pull/268/commits/5b7e26eda9934ac8869d6c47368d632a7321e741

File: src/imu/imu_sensor.cpp (hunk 5 above)
Comment (ThorsHamster) [reply src/imu/imu_sensor.cpp:14]: This is the order the content of the register is sent by the IMU. I collect the data and put them together to get the correct sensor values. It is hard to find a suitable name here.

File: src/imu/imu_sensor.cpp (hunk 7 above)
Comment (ThorsHamster) [reply src/imu/imu_sensor.cpp:61]: Thats correct. I forget to refactor this. https://github.com/Squiddrone/Squiddrone/pull/268/commits/3c6eb1d0074c2beb96c13ec248fc7fb4b589cbba

File: src/imu/imu_sensor_with_sensitivity.cpp (hunk 8 above)
Comment (ThorsHamster) [reply src/imu/imu_sensor_with_sensitivity.cpp]: This is the name in the register map of the IMU. It stands for "Full Scale Select". refactored this code, so this variable is removed.

File: tests/mock_libraries/imu/mock_i2c.hpp (hunk 11 above)
Comment (ThorsHamster) [reply tests/mock_libraries/imu/mock_i2c.hpp]: I forgot to remove this, after refactoring from tuple to pair. https://github.com/Squiddrone/Squiddrone/pull/268/commits/e1b1737d6e9ab73fd752b728d331227f6c125ccf

File: src/imu/mpu9255_data.hpp (hunk 10 above)
Comment (ThorsHamster) [reply src/imu/mpu9255_data.hpp:9]: This is the original name of the register from the register map of IMU. I could also call it "gyroscope measurements" or so, but I thought it would be easier to find it in the register map?!

File: src/imu/imu_sensor_with_sensitivity.cpp (hunk 9 above)
Comment (ThorsHamster) [reply src/imu/imu_sensor_with_sensitivity.cpp]: I refactored this code completely. I added SetBit() and ClearBit(). It would be good, to push this into some "utility" functionality later.

Comment (ThorsHamster) [reply src/imu/imu_sensor_with_sensitivity.cpp]: Refactored it in this commit: https://github.com/Squiddrone/Squiddrone/pull/268/commits/8028cb9c2bf1a1c336c0532c110af15e9c6d126f

File: src/imu/imu_sensor_with_sensitivity.cpp (hunk 8 above)
Comment (ThorsHamster) [reply src/imu/imu_sensor_with_sensitivity.cpp]: Refactored in this commit: https://github.com/Squiddrone/Squiddrone/pull/268/commits/8028cb9c2bf1a1c336c0532c110af15e9c6d126f

File: src/imu/imu_sensor.cpp (hunk 6 above)
Comment (ThorsHamster) [reply src/imu/imu_sensor.cpp]: Fixed in this commit: https://github.com/Squiddrone/Squiddrone/pull/268/commits/2ce8efcb1e7c66b36c9a258f2be564eb721664f6

File: src/imu/imu_sensor.hpp (hunk 2 above)
Comment (elheck) [reply src/imu/imu_sensor.hpp:21]: so these function should not be overwritten, correct?

File: src/imu/mpu9255_data.hpp (hunk 10 above)
Comment (elheck) [reply src/imu/mpu9255_data.hpp:9]: Ok yeah if this is a name of a register, then it is fine

File: src/imu/imu_sensor_with_sensitivity.cpp (hunk 8 above)
Comment (elheck) [reply src/imu/imu_sensor_with_sensitivity.cpp]: This is much more pleasant to read.

**Review by elheck — APPROVED** [review]
LGTM
